src/react/reconcilation.js

```diff
@@ -96,6 +103,47 @@ export type ComponentTreeState = {
   contextNodeReferences: Map<ObjectValue | AbstractObjectValue, number>,
 };
 
+function checkForSideEffects(effects: Effects, evaluatedNode: ReactEvaluatedNode): void {
+  let { modifiedBindings, result } = effects;
+  for (let [binding] of modifiedBindings) {
+    let env = binding.environment;
+    let name = binding.name;
+
+    if (env instanceof DeclarativeEnvironmentRecord) {
+      let bindings = env.bindings;
+      // ensure that this binding mutated was one that was side-effectful
+      // i.e. the binding exists from a parent scope, rather than this scope
+      if (bindings[name] === undefined || !(env instanceof FunctionEnvironmentRecord)) {
+        throw new ReconcilerRenderBailOut(
+          `Failed to optimize React component tree for "${
+            evaluatedNode.name
+          }" due to side-effects from mutating the binding "${name}"`,
+          evaluatedNode
+        );
+      }
+    } else if (env instanceof GlobalEnvironmentRecord) {
+      throw new ReconcilerRenderBailOut(
+        `Failed to optimize React component tree for "${
+          evaluatedNode.name
+        }" due to side-effects from mutating a global binding "${name}"`,
+        evaluatedNode
+      );
+    }
+  }
+  if (result instanceof PossiblyNormalCompletion || result instanceof JoinedAbruptCompletions) {
+    if (result.alternate instanceof ThrowCompletion || result.consequent instanceof ThrowCompletion) {
+      throw new ReconcilerRenderBailOut(
+        `Failed to optimize React component tree for "${
+          evaluatedNode.name
+        }" due to an exception thrown during render phase`,
+        evaluatedNode
```

Comment:
When I get this in a bundle I'm testing, there doesn't seem to be no indication of where the error happened.

The component name seems to point to the top-level component which isn't very useful. Instead, I want to know where `throw` is located, and its JS stack.

Reply:
We can probably work out from the binding, what function it was in. I'm not sure how we can make a call stack again though as this happens after all the evaluation has occured.